src/Ocelot/RateLimiting/Middleware/RateLimitingMiddleware.cs

```diff
@@ -35,6 +38,18 @@ public async Task Invoke(HttpContext httpContext)
                 return;
             }
 
+            #if NET7_0_OR_GREATER
+            if (options.RateLimitMiddlewareType == RateLimitMiddlewareType.DotNet)
+            {
+                //add EnableRateLimiting attribute to endpoint, so that .Net rate limiter can pick it up and do its thing
+                var metadata = new EndpointMetadataCollection(new EnableRateLimitingAttribute(options.RateLimiterPolicyName));
+                var endpoint = new Endpoint(null, metadata, "tempEndpoint");
```

Comment:
### WTF? 😮 
We add endpoint temporarily? Common!

Reply:
The reason I had to resort for this hack is because all the classes .Net has for this feature are marked _internal_. See [RateLimitingMiddleware](https://github.com/dotnet/aspnetcore/blob/46ea84f2c3c53c51935210d794975dcfa024ee4e/src/Middleware/RateLimiting/src/RateLimitingMiddleware.cs) and [RateLimitingMetrics](https://github.com/dotnet/aspnetcore/blob/46ea84f2c3c53c51935210d794975dcfa024ee4e/src/Middleware/RateLimiting/src/RateLimitingMetrics.cs).

This was the only way I could think of to reuse what .Net already has. I would be glad explore ff you have a suggestion for a different approach.

The `httpContext.SetEndpoint(endpoint)` call is there so that we can add `EnableRateLimitingAttribute` with the policy name for the current http request.

Reply:
I'm rephrasing my question...
**Is the setting added to the pipeline and made available for subsequent requests, or is it added only temporarily for the current request?**
----
> The reason I had to resort for this hack is because all the classes .Net has for this feature are marked internal. See [RateLimitingMiddleware](https://github.com/dotnet/aspnetcore/blob/46ea84f2c3c53c51935210d794975dcfa024ee4e/src/Middleware/RateLimiting/src/RateLimitingMiddleware.cs) and [RateLimitingMetrics](https://github.com/dotnet/aspnetcore/blob/46ea84f2c3c53c51935210d794975dcfa024ee4e/src/Middleware/RateLimiting/src/RateLimitingMetrics.cs).

OK, I see the problem! `internal` means the class can't be instantiated by constructor outside the MS lib assembly. `sealed` means the class can't be inhered by descendants. MS engineers don't wanna reshare internal implementations proposing to use only official extension methods to use the feature. This is old design dilemma for production projects when dev team don't want to process support enquiries from other developer whose projects consume the lib.
So, will you propose some else  solution if we will hack the problem of `internal sealed` qualifiers on the classes?

Reply:
> This was the only way I could think of to reuse what .Net already has. I would be glad explore if you have a suggestion for a different approach.

Currently, I haven't utilized the ASP.NET Rate Limiter. Please allow me some time to explore this feature...
As previously mentioned, anticipate several rounds of code review. My aim is to avert any issues by ensuring the feature is simple, flexible, and scalable. I'm open to coding shortcuts since our project, being an outdated ASP.NET product, already contains many. However, rest assured, we will discover a more efficient solution.
I'm confident that the injection, definition, and processing of rate limiting policies can be simplified.

Reply:
> The `httpContext.SetEndpoint(endpoint)` call is there so that we can add `EnableRateLimitingAttribute` with the policy name for the current http request.

It seems like a workaround! 😄
Okay, if I understand correctly, the issue is setting up the ASP.NET pipeline with the policy settings, and you're unsure how to do this efficiently. You're suggesting updating the pipeline and HTTP context services dynamically, correct?
Perhaps moving the setup to the application's startup phase would be better than a workaround within the middleware?
Since the policy setting is tied to DI services, any singleton/scoped DI service should suffice, or even moving the setup to the pipeline building stage, similar to [OcelotBuilder](https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot/DependencyInjection/OcelotBuilder.cs) and [BuildOcelotPipeline](https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot/Middleware/OcelotPipelineExtensions.cs), right❔

Reply:
> Is the setting added to the pipeline and made available for subsequent requests, or is it added only temporarily for the current request?

It is added temporarily. Since HttpContext is a scoped, nothing added/modified will persist into subsequent requests.

> If we consider bypassing the internal sealed restrictions, do you have an alternative solution in mind?

I don't have an alternative solution in mind at this moment.

> It seems like a workaround! 😄

It definitely is a workaround. I am trying to "trick" .Net RateLimiting middleware that it is currently executing and `Endpoint` so that the following line will not return null ([source](https://github.com/dotnet/aspnetcore/blob/46ea84f2c3c53c51935210d794975dcfa024ee4e/src/Middleware/RateLimiting/src/RateLimitingMiddleware.cs)):
`var endpoint = context.GetEndpoint();`
**and** this endpoint has `EnableRateLimiting` attribute.
`var enableRateLimitingAttribute = endpoint?.Metadata.GetMetadata<EnableRateLimitingAttribute>();`

> You're suggesting updating the pipeline and HTTP context services dynamically, correct?

Yes.

> Perhaps moving the setup to the application's startup phase would be better than a workaround within the middleware?
> Since the policy setting is tied to DI services, any singleton/scoped DI service should suffice, or even moving the setup to the pipeline building stage, similar to [OcelotBuilder](https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot/DependencyInjection/OcelotBuilder.cs) and [BuildOcelotPipeline](https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot/Middleware/OcelotPipelineExtensions.cs), right❔

I am not sure I understand this part. Are you suggesting to encapsulate following lines in a service, registering this new service in DI and invoking the service from the middleware?

 ```
var metadata = new EndpointMetadataCollection(new EnableRateLimitingAttribute(options.RateLimiterPolicyName));
var endpoint = new Endpoint(null, metadata, "tempEndpoint");
httpContext.SetEndpoint(endpoint);
```